go/vt/vtgate/tabletgateway.go

```diff
@@ -381,6 +383,109 @@ func (gw *TabletGateway) getStatsAggregator(target *querypb.Target) *TabletStatu
 	return aggr
 }
 
+// Select a random tablet out of the list of given tablets.
+//
+// Prefers tablets local to the TabletGateway's cell over tablets external to it.
+func (gw *TabletGateway) selectRandomTablet(tablets []*discovery.TabletHealth, invalidTablets map[string]bool) *discovery.TabletHealth {
+	gw.shuffleTablets(gw.localCell, tablets)
+
+	// skip tablets we tried before
+	for _, t := range tablets {
+		if _, skip := invalidTablets[topoproto.TabletAliasString(t.Tablet.Alias)]; !skip {
+			return t
+		}
+	}
+
+	return nil
+}
+
+func NewConsistentHash() *ConsistentHash {
+	return &ConsistentHash{
+		replicas: 10,
+		hashMap:  make(map[uint64]*discovery.TabletHealth),
+	}
+}
+
+type ConsistentHash struct {
+	nodes    []uint64
+	replicas int
+	hashMap  map[uint64]*discovery.TabletHealth
+}
+
+func (ch *ConsistentHash) Add(tabletHealth *discovery.TabletHealth) {
+	for i := 0; i < ch.replicas; i++ {
+		hash := xxhash.Sum64String(topoproto.TabletAliasString(tabletHealth.Tablet.Alias) + "-" + strconv.Itoa(i))
+		ch.nodes = append(ch.nodes, hash)
+		ch.hashMap[hash] = tabletHealth
+	}
+}
+
+func (ch *ConsistentHash) Sort() {
+	sort.SliceStable(ch.nodes, func(i, j int) bool { return ch.nodes[i] < ch.nodes[j] })
+}
+
+func (ch *ConsistentHash) Get(sessionUUID string) *discovery.TabletHealth {
+	hash := xxhash.Sum64String(sessionUUID)
+
+	i := sort.Search(len(ch.nodes), func(i int) bool {
+		return ch.nodes[i] >= hash
+	})
+
+	if i == len(ch.nodes) {
+		i = 0
+	}
+
+	return ch.hashMap[ch.nodes[i]]
+}
+
+// Select a random tablet out of the list of given tablets, consistently selecting the same tablet for the same session UUID.
+//
+// If the list of passed in tablets, or the list of invalid tablets changes, tablet selection will change accordingly (but stay consistent
+// barring any further changes to these two lists).
+//
+// Prefers tablets local to the TabletGateway's cell over tablets external to it.
+func (gw *TabletGateway) selectConsistentTablet(sessionUUID string, tablets []*discovery.TabletHealth, invalidTablets map[string]bool) *discovery.TabletHealth {
+	var localTablets []*discovery.TabletHealth
+	var externalTablets []*discovery.TabletHealth
+
+	for _, tablet := range tablets {
+		// Skip tablets we've tried already
+		if _, skip := invalidTablets[topoproto.TabletAliasString(tablet.Tablet.Alias)]; skip {
+			continue
+		}
+
+		if tablet.Tablet.Alias.GetCell() == gw.localCell {
+			localTablets = append(localTablets, tablet)
+		} else {
+			externalTablets = append(externalTablets, tablet)
+		}
+	}
+
+	if len(localTablets) > 0 {
+		hash := NewConsistentHash()
```

Comment:
Instead of re-generating the hashes here on each call, we could generate them whenever we get notified of a topology event (via a topology watcher) and then clone the hash struct and it's data when we need to make modifications during tablet selection.

That'd reduce the hash calls down to just the session UUID. And even that could be changed to just happen once per query.

Reply:
all should be done as part of health check subscription calls.